packages/ember-metal/lib/meta.js

```diff
@@ -0,0 +1,287 @@
+    let implementation = members[name];
+    implementation(name, Meta);
+  }
+})();
+
+
+// Implements a member that is a lazily created, non-inheritable
+// POJO.
+function ownMap(name, Meta) {
+  let key = memberProperty(name);
+  let capitalized = capitalize(name);
+  Meta.prototype['writable' + capitalized] = function() {
+    return getOrCreateOwnMap.call(this, key);
+  };
+  Meta.prototype['readable' + capitalized] = function() { return this[key]; };
+}
+
+function getOrCreateOwnMap(key) {
+  let ret = this[key];
+  if (!ret) {
+    ret = this[key] = {};
+  }
+  return ret;
+}
+
+// Implements a member that is a lazily created POJO with inheritable
+// values.
+function inheritedMap(name, Meta) {
+  let key = memberProperty(name);
+  let capitalized = capitalize(name);
+
+  Meta.prototype['writable' + capitalized] = function() {
+    return getOrCreateInheritedMap.call(this, key);
+  };
+
+  Meta.prototype['readable' + capitalized] = function() {
+    return getInherited.call(this, key);
+  };
+
+  Meta.prototype['peek' + capitalized] = function(subkey) {
+    let map = getInherited.call(this, key);
+    if (map) {
+      return map[subkey];
+    }
+  };
+
+  Meta.prototype['clear' + capitalized] = function() {
+    this[key] = {};
+  };
+}
+
+function getOrCreateInheritedMap(key) {
+  let ret = this[key];
+  if (!ret) {
+    if (this.parent) {
+      ret = this[key] = Object.create(getOrCreateInheritedMap.call(this.parent, key));
+    } else {
+      ret = this[key] = {};
+    }
+  }
+  return ret;
+}
+
+function getInherited(key) {
+  let pointer = this;
+  while (pointer) {
+    if (pointer[key]) {
+      return pointer[key];
+    }
+    pointer = pointer.parent;
+  }
+}
+
+// Implements a member that provides a lazily created map of maps,
+// with inheritance at both levels.
+function inheritedMapOfMaps(name, Meta) {
+  let key = memberProperty(name);
+  let capitalized = capitalize(name);
+
+  Meta.prototype['writable' + capitalized] = function(subkey) {
+    let outerMap = getOrCreateInheritedMap.call(this, key);
+    let innerMap = outerMap[subkey];
+    if (!innerMap) {
+      innerMap = outerMap[subkey] = {};
+    } else if (!Object.hasOwnProperty.call(outerMap, subkey)) {
+      innerMap = outerMap[subkey] = Object.create(innerMap);
+    }
+    return innerMap;
+  };
+
+  Meta.prototype['readable' + capitalized] = function(subkey) {
+    let map = getInherited.call(this, key);
+    if (map) {
+      return map[subkey];
+    }
+  };
+
+  Meta.prototype['getAll' + capitalized] = function() {
+    return getInherited.call(this, key);
+  };
+}
+
+// Implements a member that provides a non-heritable, lazily-created
+// object using the method you provide.
+function ownCustomObject(name, Meta) {
+  let key = memberProperty(name);
+  let capitalized = capitalize(name);
+  Meta.prototype['writable' + capitalized] = function(create) {
+    let ret = this[key];
+    if (!ret) {
+      ret = this[key] = create(this.source);
```

Comment:
> but you're saying it may be worth it to accept a shape-per-class in order to make instance creation faster

You know, if we do the readable/writable thing correctly, the above likely doesn't matter. As we will now only lazily create the children. In all common cases, I hope that means that we just wont incur the cost of the children at all.

We may want to add logging to detect when this assumption is violated.

We can keep it up our sleeves  if it turns out to be an issue.